### DISCUSSION TITLE
How I can a run a list of external commands from strings while gathering exit codes?

### DISCUSSION THREAD

**Discussion by Xevion** [nushell/nushell#14186] [Q&A] [answered]
I'm trying out Nushell and this stuff has genuinely got me tweaking, wondering if this language is just insanely unintuitive or if I'm not reading properly.

I'm trying to test that several directories exist, as well as that I can READ them. Unfortunately, `test -rd ($path)` doesn't work, so I have to invoke it multiple times per directory and check the status codes together.

While trying to solve this, I got stuck. My approach was this:
1. Take multiple commands in strings (no argument lists, single string per command)
2. Run each one in sequence
3. Evaluate them with `complete` (I need the exit code, and I don't want errors to propagate and kill the script).

**Note**: It is very likely I am not following best Nu practices and am posting a XY problem - if you want to suggest an alternative method (such as parsing `ls -l | get mode`) that does not involve usage of the `test` external, please do so as an _addenum_. My difficulty running external commands is valid, and the answers should reflect the title of the question (GitHub Discussions are a resource, this question will appear for others to view).

Here is the rabbit hole of development I underwent:

```nu
$ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external $in}
Error: nu::shell::eval_block_with_input

  × Eval block failed with pipeline input
   ╭─[entry #143:1:2]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external $in}
   ·  ──────────┬──────────
   ·            ╰── source value
   ╰────

Error: nu::shell::external_command

  × External command failed
   ╭─[entry #143:1:56]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external $in}
   ·                                                        ──────┬─────
   ·                                                              ╰── Command `test -d /mnt/D/DCIM` not found
   ╰────
  help: `test -d /mnt/D/DCIM` is neither a Nushell built-in or a known external command
```

Fine, I guess you want an argument list.
```nu
$ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external ($in | split row ' ')}
Error: nu::shell::eval_block_with_input

  × Eval block failed with pipeline input
   ╭─[entry #142:1:2]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external ($in | split row ' ')}
   ·  ──────────┬──────────
   ·            ╰── source value
   ╰────

Error: nu::shell::cant_convert

  × Can't convert to string.
   ╭─[entry #142:1:69]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external ($in | split row ' ')}
   ·                                                                     ──────────┬──────────
   ·                                                                               ╰── can't convert list<string> to string
```
This one through me through a loop, but I figured out it meant that `run-external` can't accept it as a list of strings. Annoying, but fine, quirk of the language.

```nu
$ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external ...($in | split row ' ')}
Error: nu::parser::missing_positional

  × Missing required positional argument.
   ╭─[entry #141:1:69]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | each {run-external ...($in | split row ' ')}
   ·                                                                     ▲
   ·                                                                     ╰── missing command
```

Fuck. Fine, I'll use a different approach.

```nu
$ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | split column ' ' | each {run-external ...$in}
Error: nu::parser::missing_positional

  × Missing required positional argument.
   ╭─[entry #145:1:88]
 1 │ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | split column ' ' | each {run-external ...$in}
   ·                                                                                        ▲
   ·                                                                                        ╰── missing command
   ╰────
  help: Usage: run-external <command> ...(args) . Use `--help` for more information.
```

Are you serious? Okay, I guess it REALLY needs to be sure that it's not the one accepting a zero length list of arguments. quirk_of_the_language.jpeg

```nu
$ ["test -d /mnt/D/DCIM", "test -r /mnt/D/DCIM"] | split column ' ' | each {run-external $in.column1 ...($in | values | skip 1 )}
╭───┬──╮
│ 0 │  │
│ 1 │  │
╰───┴──╯
```

Finally! But oh god damnit, it won't let me use `complete` or access details of the command's output. I tried looking into the `do`  command, but it doesn't let me 'ignore errors' while capturing their existence. I want to know about the errors, but I don't want them to propagate to my script.

Side note: This is a quite convoluted way to achieve `run-external`'s argument requirements. Is this _really_ the way?

**Comment by weirdan** [comment, marked as answer]
Testing several directories with `test` (assuming `test` from `coreutils`) would look like this:

```nushell
# `src` and `assets` are two directories that exist in the folder I tested this in
[src assets] | each { test -r $in -a -d $in | complete | get exit_code } | all { $in == 0 }
# outputs `true`
```

> My approach was this: Take multiple commands in strings

It goes against one of nushell's design principles: there's no `eval` in Nushell. No wonder it was hard to work it around.

**Reply by Xevion** [reply]
I did read about that design principle, but I was unaware that it extended to external commands.
Regardless, I like the composition of your answer, and that `test -r $ -a -d $` form; shell scripts aren't something I'm great at so I guess I glossed over that command syntax.

While this is a helpful answer for me personally, I still am not sure that it answers the question titled. It doesn't run external commands from a list. I re-read the details on `eval` for Nushell, and while yes it makes sense - I still don't see how it should extend to external commands.

Perhaps what I want is not possible, not because the design principles specifically state not to - but because the function I want operates too close in proximity to eval-like behavior (handling commands ran from a string), and adding more functions in that would encourage the wrong kind of thinking. Thus the feature I need doesn't exist at all. Perhaps the `run-external` command is intentionally hamstrung so you'll find an alternative solution.

**Reply by weirdan** [reply]
> It doesn't run external commands from a list. 

The answer to that is: avoid it. It's hard, it's brittle, and most of the time there are ways to achieve what you want without it.
